File: Nelson-Alvarado-Prado_Programacion-Avanzada-Trabajo#3/crud_service.py
"""
CRUD Service for Car Park Database - Python
Handles Insert, Update, Delete operations on all tables
Supports both JSON and Database sources
"""

import json
import os
from typing import Dict, Any, Optional, List, Tuple


# ================================================================
# JSON CRUD SERVICE
# ================================================================

class JsonCrudService:
    """Handle CRUD operations on JSON files"""

    def __init__(self, data_dir: str = '.'):
        self.data_dir = data_dir
        self.cars_file = os.path.join(data_dir, 'prq_cars.json')
        self.parking_file = os.path.join(data_dir, 'prq_parking.json')
        self.car_entry_file = os.path.join(data_dir, 'prq_car_entry.json')

    @staticmethod
    def _load_json(file_path: str) -> List[Dict]:
        """Load JSON file safely"""
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                return json.load(f)
        except Exception as e:
            print(f"Error loading {file_path}: {e}")
            return []

    @staticmethod
    def _save_json(file_path: str, data: List[Dict]) -> bool:
        """Save JSON file safely"""
        try:
            with open(file_path, 'w', encoding='utf-8') as f:
                json.dump(data, f, indent=2, ensure_ascii=False)
            return True
        except Exception as e:
            print(f"Error saving {file_path}: {e}")
            return False

    @staticmethod
    def _get_next_id(data: List[Dict]) -> int:
        """Get next ID for a table"""
        if not data:
            return 1
        max_id = max(item.get('id') or item.get('sequential_number') or 0 for item in data)
        return max_id + 1

    # ================================================================
    # CAR OPERATIONS
    # ================================================================

    def insert_car(self, car_data: Dict) -> Dict[str, Any]:
        """Insert a new car"""
        cars = self._load_json(self.cars_file)
        new_car = {
            'id': self._get_next_id(cars),
            'color': car_data.get('color'),
            'year': car_data.get('year'),
            'make': car_data.get('make'),
            'type': car_data.get('type')
        }
        cars.append(new_car)
        if self._save_json(self.cars_file, cars):
            return {'success': True, 'data': new_car}
        return {'success': False, 'error': 'Failed to save car'}

    def update_car(self, car_id: int, car_data: Dict) -> Dict[str, Any]:
        """Update an existing car"""
        cars = self._load_json(self.cars_file)
        car_index = next((i for i, c in enumerate(cars) if c['id'] == car_id), -1)
        if car_index == -1:
            return {'success': False, 'error': f'Car with ID {car_id} not found'}
        
        cars[car_index] = {'id': car_id, **car_data}
        if self._save_json(self.cars_file, cars):
            return {'success': True, 'data': cars[car_index]}
        return {'success': False, 'error': 'Failed to update car'}

    def delete_car(self, car_id: int) -> Dict[str, Any]:
        """Delete a car"""
        cars = self._load_json(self.cars_file)
        car_index = next((i for i, c in enumerate(cars) if c['id'] == car_id), -1)
        if car_index == -1:
            return {'success': False, 'error': f'Car with ID {car_id} not found'}
        
        deleted_car = cars[car_index]
        cars.pop(car_index)
        if self._save_json(self.cars_file, cars):
            return {'success': True, 'data': deleted_car}
        return {'success': False, 'error': 'Failed to delete car'}

    # ================================================================
    # PARKING OPERATIONS
    # ================================================================

    def insert_parking(self, parking_data: Dict) -> Dict[str, Any]:
        """Insert a new parking facility"""
        parkings = self._load_json(self.parking_file)
        new_parking = {
            'id': self._get_next_id(parkings),
            'province_name': parking_data.get('province_name'),
            'name': parking_data.get('name'),
            'hourly_rate': parking_data.get('hourly_rate')
        }
        parkings.append(new_parking)
        if self._save_json(self.parking_file, parkings):
            return {'success': True, 'data': new_parking}
        return {'success': False, 'error': 'Failed to save parking'}

    def update_parking(self, parking_id: int, parking_data: Dict) -> Dict[str, Any]:
        """Update an existing parking facility"""
        parkings = self._load_json(self.parking_file)
        parking_index = next((i for i, p in enumerate(parkings) if p['id'] == parking_id), -1)
        if parking_index == -1:
            return {'success': False, 'error': f'Parking with ID {parking_id} not found'}
        
        parkings[parking_index] = {'id': parking_id, **parking_data}
        if self._save_json(self.parking_file, parkings):
            return {'success': True, 'data': parkings[parking_index]}
        return {'success': False, 'error': 'Failed to update parking'}

    def delete_parking(self, parking_id: int) -> Dict[str, Any]:
        """Delete a parking facility"""
        parkings = self._load_json(self.parking_file)
        parking_index = next((i for i, p in enumerate(parkings) if p['id'] == parking_id), -1)
        if parking_index == -1:
            return {'success': False, 'error': f'Parking with ID {parking_id} not found'}
        
        deleted_parking = parkings[parking_index]
        parkings.pop(parking_index)
        if self._save_json(self.parking_file, parkings):
            return {'success': True, 'data': deleted_parking}
        return {'success': False, 'error': 'Failed to delete parking'}

    # ================================================================
    # CAR ENTRY OPERATIONS
    # ================================================================

    def insert_car_entry(self, entry_data: Dict) -> Dict[str, Any]:
        """Insert a new car entry"""
        entries = self._load_json(self.car_entry_file)
        new_entry = {
            'sequential_number': self._get_next_id(entries),
            'parking_id': entry_data.get('parking_id'),
            'car_id': entry_data.get('car_id'),
            'entry_date_time': entry_data.get('entry_date_time'),
            'exit_date_time': entry_data.get('exit_date_time')
        }
        entries.append(new_entry)
        if self._save_json(self.car_entry_file, entries):
            return {'success': True, 'data': new_entry}
        return {'success': False, 'error': 'Failed to save car entry'}

    def update_car_entry(self, sequential_number: int, entry_data: Dict) -> Dict[str, Any]:
        """Update an existing car entry"""
        entries = self._load_json(self.car_entry_file)
        entry_index = next((i for i, e in enumerate(entries) if e['sequential_number'] == sequential_number), -1)
        if entry_index == -1:
            return {'success': False, 'error': f'Car entry #{sequential_number} not found'}
        
        entries[entry_index] = {'sequential_number': sequential_number, **entry_data}
        if self._save_json(self.car_entry_file, entries):
            return {'success': True, 'data': entries[entry_index]}
        return {'success': False, 'error': 'Failed to update car entry'}

    def delete_car_entry(self, sequential_number: int) -> Dict[str, Any]:
        """Delete a car entry"""
        entries = self._load_json(self.car_entry_file)
        entry_index = next((i for i, e in enumerate(entries) if e['sequential_number'] == sequential_number), -1)
        if entry_index == -1:
            return {'success': False, 'error': f'Car entry #{sequential_number} not found'}
        
        deleted_entry = entries[entry_index]
        entries.pop(entry_index)
        if self._save_json(self.car_entry_file, entries):
            return {'success': True, 'data': deleted_entry}
        return {'success': False, 'error': 'Failed to delete car entry'}


# ================================================================
# DATABASE CRUD SERVICE
# ================================================================

class DatabaseCrudService:
    """Handle CRUD operations on Database"""

    def __init__(self, db_connection):
        self.db = db_connection

    # ================================================================
    # CAR OPERATIONS
    # ================================================================

    async def insert_car(self, car_data: Dict) -> Dict[str, Any]:
        """Insert a new car"""
        try:
            cursor = self.db.connection.cursor(dictionary=True)
            cursor.execute(
                'INSERT INTO PRQ_Cars (color, year, make, type) VALUES (%s, %s, %s, %s)',
                (car_data.get('color'), car_data.get('year'), car_data.get('make'), car_data.get('type'))
            )
            self.db.connection.commit()
            new_id = cursor.lastrowid
            cursor.close()
            return {'success': True, 'data': {'id': new_id, **car_data}}
        except Exception as e:
            return {'success': False, 'error': str(e)}

    async def update_car(self, car_id: int, car_data: Dict) -> Dict[str, Any]:
        """Update an existing car"""
        try:
            cursor = self.db.connection.cursor(dictionary=True)
            cursor.execute(
                'UPDATE PRQ_Cars SET color = %s, year = %s, make = %s, type = %s WHERE ID = %s',
                (car_data.get('color'), car_data.get('year'), car_data.get('make'), car_data.get('type'), car_id)
            )
            self.db.connection.commit()
            cursor.close()
            return {'success': True, 'data': {'id': car_id, **car_data}}
        except Exception as e:
            return {'success': False, 'error': str(e)}

    async def delete_car(self, car_id: int) -> Dict[str, Any]:
        """Delete a car"""
        try:
            cursor = self.db.connection.cursor(dictionary=True)
            cursor.execute('SELECT * FROM PRQ_Cars WHERE ID = %s', (car_id,))
            car = cursor.fetchone()
            if not car:
                cursor.close()
                return {'success': False, 'error': f'Car with ID {car_id} not found'}
            
            cursor.execute('DELETE FROM PRQ_Cars WHERE ID = %s', (car_id,))
            self.db.connection.commit()
            cursor.close()
            return {'success': True, 'data': car}
        except Exception as e:
            return {'success': False, 'error': str(e)}

    # ================================================================
    # PARKING OPERATIONS
    # ================================================================

    async def insert_parking(self, parking_data: Dict) -> Dict[str, Any]:
        """Insert a new parking facility"""
        try:
            cursor = self.db.connection.cursor(dictionary=True)
            cursor.execute(
                'INSERT INTO PRQ_Parking (province_name, name, hourly_rate) VALUES (%s, %s, %s)',
                (parking_data.get('province_name'), parking_data.get('name'), parking_data.get('hourly_rate'))
            )
            self.db.connection.commit()
            new_id = cursor.lastrowid
            cursor.close()
            return {'success': True, 'data': {'id': new_id, **parking_data}}
        except Exception as e:
            return {'success': False, 'error': str(e)}

    async def update_parking(self, parking_id: int, parking_data: Dict) -> Dict[str, Any]:
        """Update an existing parking facility"""
        try:
            cursor = self.db.connection.cursor(dictionary=True)
            cursor.execute(
                'UPDATE PRQ_Parking SET province_name = %s, name = %s, hourly_rate = %s WHERE ID = %s',
                (parking_data.get('province_name'), parking_data.get('name'), parking_data.get('hourly_rate'), parking_id)
            )
            self.db.connection.commit()
            cursor.close()
            return {'success': True, 'data': {'id': parking_id, **parking_data}}
        except Exception as e:
            return {'success': False, 'error': str(e)}

    async def delete_parking(self, parking_id: int) -> Dict[str, Any]:
        """Delete a parking facility"""
        try:
            cursor = self.db.connection.cursor(dictionary=True)
            cursor.execute('SELECT * FROM PRQ_Parking WHERE ID = %s', (parking_id,))
            parking = cursor.fetchone()
            if not parking:
                cursor.close()
                return {'success': False, 'error': f'Parking with ID {parking_id} not found'}
            
            cursor.execute('DELETE FROM PRQ_Parking WHERE ID = %s', (parking_id,))
            self.db.connection.commit()
            cursor.close()
            return {'success': True, 'data': parking}
        except Exception as e:
            return {'success': False, 'error': str(e)}

    # ================================================================
    # CAR ENTRY OPERATIONS
    # ================================================================

    async def insert_car_entry(self, entry_data: Dict) -> Dict[str, Any]:
        """Insert a new car entry"""
        try:
            cursor = self.db.connection.cursor(dictionary=True)
            cursor.execute(
                'INSERT INTO PRQ_Car_Entry (parking_id, car_id, entry_date_time, exit_date_time) VALUES (%s, %s, %s, %s)',
                (entry_data.get('parking_id'), entry_data.get('car_id'), entry_data.get('entry_date_time'), 
                 entry_data.get('exit_date_time'))
            )
            self.db.connection.commit()
            new_id = cursor.lastrowid
            cursor.close()
            return {'success': True, 'data': {'sequential_number': new_id, **entry_data}}
        except Exception as e:
            return {'success': False, 'error': str(e)}

    async def update_car_entry(self, sequential_number: int, entry_data: Dict) -> Dict[str, Any]:
        """Update an existing car entry"""
        try:
            cursor = self.db.connection.cursor(dictionary=True)
            cursor.execute(
                'UPDATE PRQ_Car_Entry SET parking_id = %s, car_id = %s, entry_date_time = %s, exit_date_time = %s WHERE sequential_number = %s',
                (entry_data.get('parking_id'), entry_data.get('car_id'), entry_data.get('entry_date_time'),
                 entry_data.get('exit_date_time'), sequential_number)
            )
            self.db.connection.commit()
            cursor.close()
            return {'success': True, 'data': {'sequential_number': sequential_number, **entry_data}}
        except Exception as e:
            return {'success': False, 'error': str(e)}

    async def delete_car_entry(self, sequential_number: int) -> Dict[str, Any]:
        """Delete a car entry"""
        try:
            cursor = self.db.connection.cursor(dictionary=True)
            cursor.execute('SELECT * FROM PRQ_Car_Entry WHERE sequential_number = %s', (sequential_number,))
            entry = cursor.fetchone()
            if not entry:
                cursor.close()
                return {'success': False, 'error': f'Car entry #{sequential_number} not found'}
            
            cursor.execute('DELETE FROM PRQ_Car_Entry WHERE sequential_number = %s', (sequential_number,))
            self.db.connection.commit()
            cursor.close()
            return {'success': True, 'data': entry}
        except Exception as e:
            return {'success': False, 'error': str(e)}

File: Nelson-Alvarado-Prado_Programacion-Avanzada-Trabajo#3/crud-service.js
/**
 * CRUD Service for Car Park Database - Node.js
 * Handles Insert, Update, Delete operations on all tables
 * Supports both JSON and Database sources
 */

const fs = require('fs');
const path = require('path');

// ================================================================
// JSON CRUD SERVICE
// ================================================================

class JsonCrudService {
  constructor(dataDir = '.') {
    this.dataDir = dataDir;
    this.carsFile = path.join(dataDir, 'prq_cars.json');
    this.parkingFile = path.join(dataDir, 'prq_parking.json');
    this.carEntryFile = path.join(dataDir, 'prq_car_entry.json');
  }

  /**
   * Load JSON file safely
   */
  loadJson(filePath) {
    try {
      const content = fs.readFileSync(filePath, 'utf-8');
      return JSON.parse(content);
    } catch (error) {
      console.error(`Error loading ${filePath}:`, error.message);
      return [];
    }
  }

  /**
   * Save JSON file safely
   */
  saveJson(filePath, data) {
    try {
      fs.writeFileSync(filePath, JSON.stringify(data, null, 2), 'utf-8');
      return true;
    } catch (error) {
      console.error(`Error saving ${filePath}:`, error.message);
      return false;
    }
  }

  /**
   * Get next ID for a table
   */
  getNextId(data) {
    if (data.length === 0) return 1;
    const maxId = Math.max(...data.map(item => item.id || item.sequential_number || 0));
    return maxId + 1;
  }

  // ================================================================
  // CAR OPERATIONS
  // ================================================================

  insertCar(carData) {
    const cars = this.loadJson(this.carsFile);
    const newCar = {
      id: this.getNextId(cars),
      color: carData.color,
      year: carData.year,
      make: carData.make,
      type: carData.type
    };
    cars.push(newCar);
    if (this.saveJson(this.carsFile, cars)) {
      return { success: true, data: newCar };
    }
    return { success: false, error: 'Failed to save car' };
  }

  updateCar(id, carData) {
    const cars = this.loadJson(this.carsFile);
    const carIndex = cars.findIndex(c => c.id === id);
    if (carIndex === -1) {
      return { success: false, error: `Car with ID ${id} not found` };
    }
    cars[carIndex] = { id, ...carData };
    if (this.saveJson(this.carsFile, cars)) {
      return { success: true, data: cars[carIndex] };
    }
    return { success: false, error: 'Failed to update car' };
  }

  deleteCar(id) {
    const cars = this.loadJson(this.carsFile);
    const carIndex = cars.findIndex(c => c.id === id);
    if (carIndex === -1) {
      return { success: false, error: `Car with ID ${id} not found` };
    }
    const deletedCar = cars[carIndex];
    cars.splice(carIndex, 1);
    if (this.saveJson(this.carsFile, cars)) {
      return { success: true, data: deletedCar };
    }
    return { success: false, error: 'Failed to delete car' };
  }

  // ================================================================
  // PARKING OPERATIONS
  // ================================================================

  insertParking(parkingData) {
    const parkings = this.loadJson(this.parkingFile);
    const newParking = {
      id: this.getNextId(parkings),
      province_name: parkingData.province_name,
      name: parkingData.name,
      hourly_rate: parkingData.hourly_rate
    };
    parkings.push(newParking);
    if (this.saveJson(this.parkingFile, parkings)) {
      return { success: true, data: newParking };
    }
    return { success: false, error: 'Failed to save parking' };
  }

  updateParking(id, parkingData) {
    const parkings = this.loadJson(this.parkingFile);
    const parkingIndex = parkings.findIndex(p => p.id === id);
    if (parkingIndex === -1) {
      return { success: false, error: `Parking with ID ${id} not found` };
    }
    parkings[parkingIndex] = { id, ...parkingData };
    if (this.saveJson(this.parkingFile, parkings)) {
      return { success: true, data: parkings[parkingIndex] };
    }
    return { success: false, error: 'Failed to update parking' };
  }

  deleteParking(id) {
    const parkings = this.loadJson(this.parkingFile);
    const parkingIndex = parkings.findIndex(p => p.id === id);
    if (parkingIndex === -1) {
      return { success: false, error: `Parking with ID ${id} not found` };
    }
    const deletedParking = parkings[parkingIndex];
    parkings.splice(parkingIndex, 1);
    if (this.saveJson(this.parkingFile, parkings)) {
      return { success: true, data: deletedParking };
    }
    return { success: false, error: 'Failed to delete parking' };
  }

  // ================================================================
  // CAR ENTRY OPERATIONS
  // ================================================================

  insertCarEntry(entryData) {
    const entries = this.loadJson(this.carEntryFile);
    const newEntry = {
      sequential_number: this.getNextId(entries),
      parking_id: entryData.parking_id,
      car_id: entryData.car_id,
      entry_date_time: entryData.entry_date_time,
      exit_date_time: entryData.exit_date_time || null
    };
    entries.push(newEntry);
    if (this.saveJson(this.carEntryFile, entries)) {
      return { success: true, data: newEntry };
    }
    return { success: false, error: 'Failed to save car entry' };
  }

  updateCarEntry(sequentialNumber, entryData) {
    const entries = this.loadJson(this.carEntryFile);
    const entryIndex = entries.findIndex(e => e.sequential_number === sequentialNumber);
    if (entryIndex === -1) {
      return { success: false, error: `Car entry #${sequentialNumber} not found` };
    }
    entries[entryIndex] = { sequential_number: sequentialNumber, ...entryData };
    if (this.saveJson(this.carEntryFile, entries)) {
      return { success: true, data: entries[entryIndex] };
    }
    return { success: false, error: 'Failed to update car entry' };
  }

  deleteCarEntry(sequentialNumber) {
    const entries = this.loadJson(this.carEntryFile);
    const entryIndex = entries.findIndex(e => e.sequential_number === sequentialNumber);
    if (entryIndex === -1) {
      return { success: false, error: `Car entry #${sequentialNumber} not found` };
    }
    const deletedEntry = entries[entryIndex];
    entries.splice(entryIndex, 1);
    if (this.saveJson(this.carEntryFile, entries)) {
      return { success: true, data: deletedEntry };
    }
    return { success: false, error: 'Failed to delete car entry' };
  }
}

// ================================================================
// DATABASE CRUD SERVICE
// ================================================================

class DatabaseCrudService {
  constructor(pool) {
    this.pool = pool;
  }

  // ================================================================
  // CAR OPERATIONS
  // ================================================================

  async insertCar(carData) {
    const connection = await this.pool.getConnection();
    try {
      const [result] = await connection.execute(
        'INSERT INTO PRQ_Cars (color, year, make, type) VALUES (?, ?, ?, ?)',
        [carData.color, carData.year, carData.make, carData.type]
      );
      return { success: true, data: { id: result.insertId, ...carData } };
    } catch (error) {
      return { success: false, error: error.message };
    } finally {
      connection.release();
    }
  }

  async updateCar(id, carData) {
    const connection = await this.pool.getConnection();
    try {
      await connection.execute(
        'UPDATE PRQ_Cars SET color = ?, year = ?, make = ?, type = ? WHERE ID = ?',
        [carData.color, carData.year, carData.make, carData.type, id]
      );
      return { success: true, data: { id, ...carData } };
    } catch (error) {
      return { success: false, error: error.message };
    } finally {
      connection.release();
    }
  }

  async deleteCar(id) {
    const connection = await this.pool.getConnection();
    try {
      const [rows] = await connection.execute('SELECT * FROM PRQ_Cars WHERE ID = ?', [id]);
      if (rows.length === 0) {
        return { success: false, error: `Car with ID ${id} not found` };
      }
      const car = rows[0];
      await connection.execute('DELETE FROM PRQ_Cars WHERE ID = ?', [id]);
      return { success: true, data: car };
    } catch (error) {
      return { success: false, error: error.message };
    } finally {
      connection.release();
    }
  }

  // ================================================================
  // PARKING OPERATIONS
  // ================================================================

  async insertParking(parkingData) {
    const connection = await this.pool.getConnection();
    try {
      const [result] = await connection.execute(
        'INSERT INTO PRQ_Parking (province_name, name, hourly_rate) VALUES (?, ?, ?)',
        [parkingData.province_name, parkingData.name, parkingData.hourly_rate]
      );
      return { success: true, data: { id: result.insertId, ...parkingData } };
    } catch (error) {
      return { success: false, error: error.message };
    } finally {
      connection.release();
    }
  }

  async updateParking(id, parkingData) {
    const connection = await this.pool.getConnection();
    try {
      await connection.execute(
        'UPDATE PRQ_Parking SET province_name = ?, name = ?, hourly_rate = ? WHERE ID = ?',
        [parkingData.province_name, parkingData.name, parkingData.hourly_rate, id]
      );
      return { success: true, data: { id, ...parkingData } };
    } catch (error) {
      return { success: false, error: error.message };
    } finally {
      connection.release();
    }
  }

  async deleteParking(id) {
    const connection = await this.pool.getConnection();
    try {
      const [rows] = await connection.execute('SELECT * FROM PRQ_Parking WHERE ID = ?', [id]);
      if (rows.length === 0) {
        return { success: false, error: `Parking with ID ${id} not found` };
      }
      const parking = rows[0];
      await connection.execute('DELETE FROM PRQ_Parking WHERE ID = ?', [id]);
      return { success: true, data: parking };
    } catch (error) {
      return { success: false, error: error.message };
    } finally {
      connection.release();
    }
  }

  // ================================================================
  // CAR ENTRY OPERATIONS
  // ================================================================

  async insertCarEntry(entryData) {
    const connection = await this.pool.getConnection();
    try {
      const [result] = await connection.execute(
        'INSERT INTO PRQ_Car_Entry (parking_id, car_id, entry_date_time, exit_date_time) VALUES (?, ?, ?, ?)',
        [entryData.parking_id, entryData.car_id, entryData.entry_date_time, entryData.exit_date_time || null]
      );
      return { success: true, data: { sequential_number: result.insertId, ...entryData } };
    } catch (error) {
      return { success: false, error: error.message };
    } finally {
      connection.release();
    }
  }

  async updateCarEntry(sequentialNumber, entryData) {
    const connection = await this.pool.getConnection();
    try {
      await connection.execute(
        'UPDATE PRQ_Car_Entry SET parking_id = ?, car_id = ?, entry_date_time = ?, exit_date_time = ? WHERE sequential_number = ?',
        [entryData.parking_id, entryData.car_id, entryData.entry_date_time, entryData.exit_date_time || null, sequentialNumber]
      );
      return { success: true, data: { sequential_number: sequentialNumber, ...entryData } };
    } catch (error) {
      return { success: false, error: error.message };
    } finally {
      connection.release();
    }
  }

  async deleteCarEntry(sequentialNumber) {
    const connection = await this.pool.getConnection();
    try {
      const [rows] = await connection.execute(
        'SELECT * FROM PRQ_Car_Entry WHERE sequential_number = ?',
        [sequentialNumber]
      );
      if (rows.length === 0) {
        return { success: false, error: `Car entry #${sequentialNumber} not found` };
      }
      const entry = rows[0];
      await connection.execute('DELETE FROM PRQ_Car_Entry WHERE sequential_number = ?', [sequentialNumber]);
      return { success: true, data: entry };
    } catch (error) {
      return { success: false, error: error.message };
    } finally {
      connection.release();
    }
  }
}

module.exports = {
  JsonCrudService,
  DatabaseCrudService
};
